### DISCUSSION TITLE
False Positive | cuda.OutOfMemoryError | batch size | General Training Help

### DISCUSSION THREAD

**Discussion by AzwadFawadHasan** [ultralytics/ultralytics#2259] [Q&A] [answered]
Hi, I am new to YOLO. I am trying to train a yolov8x.pt model to correctly detect 9 different faces. I have used an iphoneX to take pictures of the faces and then uploaded them on roboflow.com to label them. My training, validation and testing set contains 5000 images, 161, 147 images respectively. I have used Smart Polygon Enhanced to label most of the images and used bounding boxes to label the rest. Each of the different face classes contains 230-250 images on average. Even after labeling the images properly and training them multiple times I am getting several false positives, sometimes the model detects curtains and random objects as faces.

I have used the the CLI command below to train the weights
yolo detect train data=mypath/data.yaml model=mypath/runs/train3/last.pt pretrained=mypath/runs/train3/best.pt  epochs=100 cache=True batch=-1 imgsz=640

# Question no1
I am using a Laptop with R7 5700H cpu, 3050ti 4GB GPU, 32GB RAM. Cuda error shows up whenever I use batches larger than 3. When I use batch=-1 It goes through all the images during each epochs. Is there anyway I can fix this issue? Will using a larger batch size give me better results

cuda.outOfMemoryError
![image](https://user-images.githubusercontent.com/106096161/234591908-46b8f53b-94d5-4a78-a370-24e73504748f.png)

# Question no2.
While training using batch=-1, the task manager shows 0percent utilization of my gpu. Is my GPU being used to train the model? Also, when I add cache=True during training the it/s for each epochs increases to 6.5 it/s.  If I don't the rate falls drastically. 

works when batch size =2 but it/s is very less and shows gpu usage is 0
![image](https://user-images.githubusercontent.com/106096161/234594113-0b0aaeb0-818b-4f9c-ad8a-a6ad6e9e6a7a.png)
![image](https://user-images.githubusercontent.com/106096161/234594297-c16d2df1-ceff-4005-8567-062c537225cf.png)

# Question no3
Since I used  Roboflow's Smart Polygon Enhanced to label the images, this message appears during training/ Is this affecting my results by anychance?. Should I re-do all of the annotations to bounding boxes?
roboflow sam
![image](https://user-images.githubusercontent.com/106096161/234593147-eeac10dc-9501-4ae3-94c8-01a76b2548ac.png)

# Question no4
What's the correct way to train a model to get a good result? What hyper parameter's should I use/change to get better face recognition?

# Question no5
What's the difference between these two CLI commands?
//Start training from a pretrained *.pt model
yolo detect train data=mypath/data.yaml model=mypath/runs/train3/last.pt epochs=100 imgsz=640 

//Build a new model from YAML, transfer pretrained weights to it and start training
yolo detect train data=mypath/data.yaml model=mypath/runs/train3/last.pt pretrained=mypath/runs/train3/best.pt epochs=100 imgsz=640

Should I create separate dataset for X-face and Y-face , train models separately and then transfer the pretrained models to retrain them again with Z-face? Will this help me get better results?

**Comment by AzwadFawadHasan** [comment]
@glenn-jocher  Hii, could you please help

**Reply by glenn-jocher** [reply, marked as answer]
@AzwadFawadHasan hello! Let's address your questions one by one:

1. **CUDA Out of Memory Error**: The `cuda.OutOfMemoryError` indicates that your GPU's memory is insufficient for the batch size you're trying to use. Reducing the batch size is the right approach here. Larger batch sizes can sometimes improve results, but they are primarily for faster training. With your GPU, sticking to smaller batch sizes is necessary. Using gradient accumulation might help you simulate larger batch sizes without running into memory issues.

2. **GPU Utilization**: If Task Manager shows 0% GPU utilization, it's possible that your GPU is not being used for training. Ensure that you have the correct CUDA drivers and PyTorch version installed. The `cache=True` parameter speeds up training by caching images in RAM after the first epoch, which explains the increase in iterations per second (it/s).

3. **Annotations**: The message regarding Roboflow's Smart Polygon Enhanced annotations is likely informational and should not affect your results unless the format is incompatible with YOLOv8. Ensure that the annotations are correctly converted to a format YOLOv8 can understand.

4. **Training for Better Results**: To improve face recognition, experiment with different hyperparameters such as learning rate, weight decay, and augmentation settings. Also, ensure that your dataset is diverse and representative of the faces you want to detect. Overfitting to a particular set of images can lead to poor generalization.

5. **CLI Commands Difference**: The first command starts training from a pretrained model, while the second command also specifies a `pretrained` path, which is redundant if you're already specifying the `model` path. The `pretrained` argument is typically used to specify a path to pretrained weights that you want to transfer to a new model.

6. **Separate Datasets for Different Faces**: Training separate models for different faces and then combining them is not typically necessary. YOLOv8 should be able to handle multiple classes in a single model. Ensure that your dataset for each class is balanced and that you're not overfitting to a particular class.

For further guidance on training and hyperparameter tuning, please refer to the Ultralytics Docs. Remember, achieving the best results often requires iterative experimentation and patience. Good luck with your training! 🚀

**Reply by AzwadFawadHasan** [reply]
@glenn-jocher  Thanks for getting back to me. After diving deeper into YOLO, I've noticed it might not be the best choice for recognizing faces compared to other models designed specifically for that purpose. Still, the tips shared here should help beginners make the most out of their YOLO training for better results.

**Reply by glenn-jocher** [reply]
You're welcome, @AzwadFawadHasan! It's true that while YOLO is a powerful object detection system, it may not always be the optimal choice for face recognition tasks, especially when compared to models specifically tailored for facial feature extraction and recognition. For face recognition, you might want to explore models like FaceNet or DeepFace, which are designed with facial biometrics in mind.

However, if you're set on using YOLO for face detection as a precursor to recognition, ensuring high-quality, diverse training data and fine-tuning your model can still yield good results. If you have any more questions or need further assistance with YOLO or other models, feel free to reach out. Happy to help, and best of luck with your project! 😊👍
